R/get-path-from-object.R

```diff
@@ -158,32 +158,188 @@ get_model_id.bbi_model <- function(.mod) {
 #' @export
-get_data_path.bbi_model <- function(.mod, ...) {
+get_data_path.bbi_nonmem_model <- function(
+    .bbi_object,
+    .check_exists = TRUE,
+    pull_from_config = FALSE,
+    ...
+){
+  get_data_path_nonmem(.bbi_object, .check_exists, pull_from_config, ...)
+}
+
+#' @rdname get_data_path
+#' @export
+get_data_path.bbi_nonmem_summary <- function(
+    .bbi_object,
+    .check_exists = TRUE,
+    pull_from_config = FALSE,
+    ...
+){
+  get_data_path_nonmem(.bbi_object, .check_exists, pull_from_config, ...)
+}
+
+#' @rdname get_data_path
+#' @export
+get_data_path.bbi_log_df <- function(
+    .bbi_object,
+    .check_exists = TRUE,
+    pull_from_config = FALSE,
+    ...
+){
+  data_paths <- map_chr(.bbi_object[[ABS_MOD_PATH]], function(.path) {
+    .mod <- read_model(.path)
+    get_data_path_nonmem(.mod, .check_exists, pull_from_config, ...)
+  })
+
+  return(data_paths)
+}
+
+
+#' @keywords internal
+get_data_path_nonmem <- function(
+    .bbi_object,
+    .check_exists = TRUE,
+    pull_from_config = FALSE,
+    ...
+){
+
+  if(inherits(.bbi_object, NM_SUM_CLASS)){
+    .bbi_object <- read_model(.bbi_object$absolute_model_path)
+  }
+  check_model_object(.bbi_object, "bbi_nonmem_model")
+  mod_path <- get_model_path(.bbi_object)
+
+  # Get relative data path
+  data_path_rel <- get_data_path_from_ctl(.bbi_object)
+
+  # Handling for `.mod` extensions
+  if(grepl("(?i)ctl", fs::path_ext(mod_path))){
+    path_elements <- unlist(strsplit(data_path_rel, "/"))
+    if (path_elements[1] == "..") {
+      data_path_rel_adj <- paste(path_elements[-1], collapse = "/")
+    }else{
+      data_path_rel_adj <- data_path_rel
+    }
+  }else{
+    data_path_rel_adj <- data_path_rel
+  }
+
+  data_path <- fs::path_norm(file.path(dirname(mod_path), data_path_rel_adj)) %>%
+    as.character()
+
+  if(.check_exists){
+    if(!fs::file_exists(data_path)){
+      msg_extra <- NULL
+      # Check for improper path specification due to extensions
+      if(grepl("(?i)ctl", fs::path_ext(mod_path))){
+        data_path_test <- fs::path_norm(file.path(dirname(mod_path), data_path_rel)) %>%
+          as.character()
+        # If data file exists here, this means the user was missing an extra `../`
+        # for `ctl` extensions. Give informative message in this case.
+        if(fs::file_exists(data_path_test)){
+          msg_extra <- c(
+            "i" = paste(
+              "Your model ends with a `.ctl` extension - make sure",
+              "you include an extra `../` in the relative path.",
+              "See `?get_data_path` for details..."
+            )
+          )
+        }
+      }
+
+      # The first error message line is what NMTRAN would return in this situation
+      rlang::abort(
+        c(
+          "x" = "Input data file does not exist or cannot be opened",
+          "i" = glue("Referenced input data path: {data_path_rel}"),
+          msg_extra
+        )
+      )
+    }
+  }
+
+  if(isTRUE(pull_from_config)){
```

Comment:
I don't understand the rationale for doing both of these and comparing, rather than just choosing one code path based on whether `pull_from_config = TRUE`.

Reply:
This was suggested by @seth127 [here](https://github.com/metrumresearchgroup/bbr/issues/651#issuecomment-1942423377):
>  If a bbi_config.json does exist, check the path and warn if they don't match

Reply:
> This was suggested by @seth127 [here](https://github.com/metrumresearchgroup/bbr/issues/651#issuecomment-1942423377):

That's fine, but "@seth127 said to do it" isn't the rationale that, as a reviewer, I'm trying to understand and get you to document/expand on.

---

My current view is that the approach here is overcomplicating things.  If the caller asked to pull from the config, get it from there and error if you can't.   If they said don't get or from the config, then get it from the model.

(That's taking for granted the current design of having an argument.  I'll expand on it later if I don't talk myself out of it while writing the comment, but one thing I wonder is whether we should just have the `bbi_model` method retain the bbi config handling and then have a custom `bbi_nonmem_model` method that falls back to extracting it from the control stream if the config file doesn't exist.  The `bbi_nonmem_model` method should of course document that custom handling on top.  Anyway, as I said, haven't though that through yet, and there are of course other variations.)

Reply:
I dont really have any opinion on this, which is why I pointed you to Seth. Im fine with either direction. That being said, in my most recent refactor I did remove this check and followed your recommendation about exclusively pulling from one location or the other.

Reply:
It looks like that point in the [specification](https://github.com/metrumresearchgroup/bbr/issues/651#issuecomment-1942423377) was probably prompted by this part of your initial post in gh-651:

> I propose we attempt to fetch _both_ data paths. i.e. we extract the data path from the `ctl` file, and _if_ the model has been run, we also parse it from the `bbr` model object. We could compare the two, and warn the user if there is a mismatch (i.e. if the model file has been moved since it was run). We could have an argument for which one to return (with a default), to ensure a single string is usually returned.

But part of the reason why I'm confused about the code I was commenting on is that, as implemented, the check is done only with the non-default `pull_from_config = TRUE`.  That doesn't seem to align with the motivation described above.

In particular, the "control stream data path no longer matches with previous run" scenarios I see are:

 1. the caller goes with the default `pull_from_config = FALSE`. They are not warned about the mismatch.

 2. the caller passes an explicit `pull_from_config = FALSE`. They are not warned about mismatch.

 3. the caller passes a non-default `pull_from_config = TRUE`. They are warned about the mismatch.

Of all of these, the first seems like the one that it'd be most where it'd be most important to warn (because the caller isn't explicitly asking for source), but the check isn't done in that case.

I know the warning is now removed entirely. To me, whether that is a good direction to go depends how much we're worried about a scenario like this:

 * caller runs model

 * caller then edits control stream and points it to a different data path

 * caller later calls `nm_join()`. In this case, the appropriate data to join is the path in the config. That is was what was fed to the model that produced the result.

If we pull from the control stream by default, the above seems like an important mistake to guard against even if it will be very rare. (And again, the code I was commenting would not have warned here.)

Stepping back a bit, shouldn't `nm_join()` (or `nm_tables()` underneath, I suppose) explicitly request the config data here? It's concerned with an output run, which is tied to the data in the config regardless of what the current value in the control stream is.